piet-coregraphics/src/text.rs

```diff
@@ -0,0 +1,433 @@
+        let lines = self.unwrap_frame().get_lines();
+        let line = lines.get(line_number.min(isize::max_value() as usize) as isize)?;
+        let line = Line::new(&line);
+        let typo_bounds = line.get_typographic_bounds();
+        let (start_offset, end_offset) = self.line_range(line_number)?;
+        let text = self.line_text(line_number)?;
+        //FIXME: this is just ascii whitespace
+        let trailing_whitespace = text
+            .as_bytes()
+            .iter()
+            .rev()
+            .take_while(|b| match b {
+                b' ' | b'\t' | b'\n' | b'\r' => true,
+                _ => false,
+            })
+            .count();
+        let height = typo_bounds.ascent + typo_bounds.descent + typo_bounds.leading;
+        // this may not be exactly right, but i'm also not sure we ever use this?
+        //  see https://stackoverflow.com/questions/5511830/how-does-line-spacing-work-in-core-text-and-why-is-it-different-from-nslayoutm
+        let cumulative_height =
+            (self.line_y_positions[line_number] + typo_bounds.descent + typo_bounds.leading).ceil();
+        Some(LineMetric {
+            start_offset,
+            end_offset,
+            trailing_whitespace,
+            baseline: typo_bounds.ascent,
+            height,
+            cumulative_height,
+        })
+    }
+
+    fn line_count(&self) -> usize {
+        self.line_y_positions.len()
+    }
+
+    // given a point on the screen, return an offset in the text, basically
+    fn hit_test_point(&self, point: Point) -> HitTestPoint {
+        let mut line_num = self
+            .line_y_positions
+            .iter()
+            .position(|y| y >= &point.y)
+            // if we're past the last line, use the last line
+            .unwrap_or_else(|| self.line_y_positions.len().saturating_sub(1));
+        // because y_positions is the position of the baseline, check that we don't
+        // fall between the preceding baseline and that line's descent
+        if line_num > 0 {
+            let prev_line = self.unwrap_frame().get_line(line_num - 1).unwrap();
+            let typo_bounds = Line::new(&&prev_line).get_typographic_bounds();
+            if self.line_y_positions[line_num - 1] + typo_bounds.descent >= point.y {
+                line_num -= 1;
+            }
+        }
+        let line: Line = self
+            .unwrap_frame()
+            .get_line(line_num)
+            .map(Into::into)
+            .unwrap();
+        let fake_y = self.line_y_positions[line_num];
+        // map that back into our inverted coordinate space
+        let fake_y = -(self.frame_size.height - fake_y);
+        let point_in_string_space = CGPoint::new(point.x, fake_y);
+        let offset_utf16 = line.get_string_index_for_position(point_in_string_space);
+        let offset = match offset_utf16 {
+            // this is 'kCFNotFound'.
+            // if nothing is found just go end of string? should this be len - 1? do we have an
```

Comment:
Gut feeling is that `len` is reasonable. `len - 1` is almost certainly wrong as it might cut a codepoint.